File: [H]array/[H]array-prefix-sum/238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1 for _ in range(len(nums))]

        total = 1
        for i, num in enumerate(nums):
            res[i] = total
            total *= num
        
        total = 1
        for i in range(len(nums) - 1, - 1, - 1):
            res[i] *= total
            total *= nums[i]

        return res
    
# time O(n)
# space O(1), not counting output list
# using array and prefix sum and standard prefix sum

File: [H]array/[H]array-prefix-sum/238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] res = new int[nums.length];
        Arrays.fill(res, 1);
        int total = 1;
        for (int i = 0; i < nums.length; i++) {
            res[i] = total;
            total *= nums[i];
        }
        total = 1;
        for (int i = nums.length - 1; i >= 0; i--) {
            res[i] *= total;
            total *= nums[i];
        }
        return res;
    }
}

// time O(n)
// space O(1), not counting output list
// using array and prefix sum and standard prefix sum